### PR TITLE
MobsV2 - Part 1

### PR DESCRIPTION
**Pull request by MaxIsJoe**
Improved AI handling for mobs
NPCs that are on V2 now are fully compatible with HealthV2 features

This introduction to MobV2 is basically done, with a few quirks that need to be ironed out in a later date.
More mobs should be ported over in the upcoming PRs alongside their AI, but for now CowV2 is the only example for this.

Soon, we'll be able to finally nuke old deprecated mob and health code completely out of the project

### CODE REVIEW

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public abstract class LivingHealthMasterBase : NetworkBehaviour, IFireExposable,
 		/// <summary>
 		/// Rate at which periodic damage, such as radiation, should be applied
 		/// </summary>
-		private float tickRate = 1f;
+		[SerializeField] private float tickRate = 1f;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:55]: oh?

File: UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs [hunk 2]
```diff
@@ -33,6 +33,7 @@ public class ReagentPoolSystem : HealthSystemBase
 		public void AddFreshBlood(ReagentMix bloodPool, float amount)
 		{
 			// Currently only does blood and required reagents, should at nutriments and other common gases
+			if(bloodPool == null || bloodType == null || bloodType.CirculatedReagent == null) return;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs]: This is because you haven't it setup initial system on the mob

File: UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+
+namespace Mobs.AI
+{
+	public class MobAI : MonoBehaviour
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: there is now ready system for MobAI? 
with the contemplate goal and stuff

File: UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs]: Well I don't want to, but It's trying to set it up by itself even though I did not add any systems to it.
so the quickest solution to me was to just check if anything was null and not block the entire foreach loop for initializing because one rouge system

File: UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: Someone removed it sometime ago, and I couldn't find it again. So I took the simplest approach to finte state machine, we don't need the fancy visuals; just a gameObject to list all states and a way to keep track of all the active ones.

File: UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs]: is because the brain has  blood circulated component, and so that Generates the system

File: UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs (hunk 3 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: Public class MobController : MonoBehaviour
it still there

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: it will be deleted when we move all mobs to V2

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: The question is which is better?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: mine

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: Need discussion with everyone,  To make sure it's okay

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 1 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:55]: why is this opened to the inspector?

File: UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs (hunk 3 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Mobs/AI/MobAI.cs:9]: yeah, had discussiono n discord moving to this system for AI

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:55]: mobs don't need to be as actively checked as players

Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:55]: might be better to do it if a player is not possessing it?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:55]: or you can just expose the field and leave it as is 
players possessing other NPCs aren't high on the list right now to warrent adding a way to change the tickrate based on if a player is possessing or not.

it also makes the behavior much more consistent regardless of what is controlling the mob

Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:55]: but gives inconsistent behaviour if you put a body part from a mob into a player or vice versa , I'm fine if you  just slap a TODO on it

File: UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/PolymorphicSystems/ReagentPoolSystem.cs]: removed the return and added an error to remind people about defining systems

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+				{
+					isComplete = true;
+					return null;
+				}
+			}
+
+			return null;
+		}
+
+		void FindNeighbours(Node currentNode)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:194]: You could use the meta data system Neighbours Use by atmospherics

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+		protected virtual void OnTileReached(Vector3Int oldLocalPos, Vector3Int newLocalPos)
+		{
+			if (movingToTile && !arrivedAtTile)
+			{
+				arrivedAtTile = true;
+			}
+		}
+
+		private void RefreshNodeType(Node node)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:402]: why is this here?

File: UnityProject/Assets/Scripts/Player/PlayerSprites.cs [hunk 6]
```diff
@@ -432,34 +433,33 @@ private void UpdateBurningOverlays(float fireStacks, OrientationEnum currentFaci
...
+			ThisCharacter = characterSettings;
+			bool notCustomSpecies = string.IsNullOrEmpty(RaceOverride);
+			RaceBodyparts = gameObject.TryGetComponent<Mob>(out var mob) ?
+				characterSettings.GetRaceSoNoValidation() : characterSettings.GetRaceSo(notCustomSpecies);
 
-				livingHealthMasterBase.InitialiseFromRaceData(RaceBodyparts);
-				livingHealthMasterBase.SetUpCharacter(RaceBodyparts);
-				SetupSprites();
-				livingHealthMasterBase.StartFresh();
+			if (RaceBodyparts == null)
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Player/PlayerSprites.cs:454]: this shouldn't run if it's null, that shouldn't also auto populate with human, since aliens requires the null

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:402]: idk

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:194]: I'm planning on updating the pathfinder in later PRs, but for now it uses the already existing code

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs (hunk 5 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:402]: isn't it meant to be in the meta data node system?

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:194]: as in copy and paste from the old pathfinder?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:194]: for now it is, yes.
it only has some elements updated to use things like the new UOP stuff

File: UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Mobs/AI/MobPathfinderV2.cs:402]:
```asm
include io.h
data segment
  character db 'idk',0
  data ends
code segment
  start :
  repnz db
  jmp start 
data ends
end start
```

**Review by Bod9001 — APPROVED** [review]
